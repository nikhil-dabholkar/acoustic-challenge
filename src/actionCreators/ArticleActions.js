import { LOAD_ARTICLE_DETAILS_DATA, LOAD_ALL_ARTICLES_DATA } from '../constants/constants';
const axios = require('axios');

export const fetchSelectedArticleData = (articleId) => {
    return async (dispatch) => {
        const response = await axios.get(`https://content-eu-4.content-cms.com/api/859f2008-a40a-4b92-afd0-24bb44d10124/delivery/v1/content/${articleId}`);
        console.log(response);
        dispatch({
            type: LOAD_ARTICLE_DETAILS_DATA,
            payload: response.data
        })
    }
}

export const fetchAllArticlesData = () => {
    return async (dispatch) => {
        const response = await axios.get(`https://run.mocky.io/v3/8421cb68-a9d9-4575-ad5d-b4396c864231/`);
        console.log(response);
        dispatch({
            type: LOAD_ALL_ARTICLES_DATA,
            payload: response.data
        })
    }
}
