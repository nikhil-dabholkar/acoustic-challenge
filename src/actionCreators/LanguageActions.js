import { SELECT_LANGUAGE_PREFERENCE } from '../constants/constants'

/**
 * This action will be called when user changes the language
 */
export const switchLanguagePreference = (payload) => ({
  type: SELECT_LANGUAGE_PREFERENCE,
  payload: payload
})
