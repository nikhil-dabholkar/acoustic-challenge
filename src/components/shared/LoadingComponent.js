import React from 'react';
import {Loading} from 'carbon-components-react';


function LoadingComponent() {
    return (<React.Fragment>
                <Loading description="Active loading indicator" withOverlay={true} />
            </React.Fragment>
    );
}

export default LoadingComponent;
