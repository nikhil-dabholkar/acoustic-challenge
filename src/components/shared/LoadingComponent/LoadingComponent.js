import React from 'react'
import { Loading } from 'carbon-components-react'

function LoadingComponent () {
  return (<React.Fragment>
    <div data-testid="loadingComponent">
      <Loading description="Active loading indicator" withOverlay={true} />
    </div>
  </React.Fragment>
  )
}

export default LoadingComponent
