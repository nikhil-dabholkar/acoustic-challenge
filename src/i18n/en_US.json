{
    "app.header": "English",
    "detailsComponent.dateTooltip": "Date and time is used when ordering a list of design articles. The time will not display in the website.",
    "detailsComponent.bodySectionTooltip": "The design topic, when displayed in the website, will interleave the text chunks and images to create a fluid layout.",
    "detailsComponent.topicTitle": "Topic title",
    "detailsComponent.authorName": "Author name",
    "detailsComponent.date": "Date",
    "detailsComponent.leadImage": "Lead image",
    "detailsComponent.leadImageCaption": "Lead image caption",
    "detailsComponent.LeadImageCredit": "Lead image credit",
    "detailsComponent.profiles": "Profiles",
    "detailsComponent.lead": "Lead",
    "detailsComponent.card": "Card",
    "detailsComponent.default": "Default",
    "detailsComponent.textForBodySection": "Text for body section",
    "listComponent.id": "Id",
    "listComponent.name": "Name",
    "headerComponent.home": "Home",
    "headerComponent.articles": "Articles"
}