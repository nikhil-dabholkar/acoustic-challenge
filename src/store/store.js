import rootReducer from '../reducer/index';
import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';

export const store = createStore(rootReducer, applyMiddleware(thunk))