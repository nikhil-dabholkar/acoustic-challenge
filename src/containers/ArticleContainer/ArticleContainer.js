import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchSelectedArticleData, fetchAllArticlesData, resetArticlesError } from '../../actionCreators/ArticleActions';
import { Route, useLocation } from "react-router-dom";
import { useHistory } from 'react-router-dom'
import { lazy, Suspense } from 'react';
import { LoadingComponent } from '../../components/index';
import { ErrorBoundaryWrapper } from '../../wrappers/index';

const ListComponent = lazy(() => import("../../components/article/ListComponent/ListComponent"));
const DetailsComponent = lazy(() => import("../../components/article/DetailsComponent/DetailsComponent"));

function ArticleContainer({ resetError, fetchArticle, fetchAllArticles, article }) {
    const history = useHistory();
    const location = useLocation();
    const errorBoundaryRef = useRef(null);

    useEffect(() => {
        /**
         * When ever the url changes, clear the error 
         */
        return history.listen((location) => {
            resetError();
        })
    }, [history])

    return (<React.Fragment>
        <ErrorBoundaryWrapper key={location.pathname} ref={errorBoundaryRef}>
            <div data-testid="articleContainer">
                <Route path="/articles" render={(props) => {
                    return (<Suspense fallback={<LoadingComponent />}>
                        <ListComponent loading={article.loading} error={article.error} allArticles={article.allArticles} fetchAllArticles={fetchAllArticles} {...props} />
                    </Suspense>);
                }} exact />
                <Route path="/articles/:id" render={(props) => {
                    return (<Suspense fallback={<LoadingComponent />}>
                        <DetailsComponent loading={article.loading} error={article.error} selectedArticle={article.selectedArticle} fetchArticle={fetchArticle} {...props} />
                    </Suspense>)
                }} exact />
            </div>
        </ErrorBoundaryWrapper>
    </React.Fragment>
    );
}

const mapStateToProps = state => {
    return ({
        article: state.articleReducer
    })
};

const mapDispatchToProps = {
    fetchArticle: fetchSelectedArticleData,
    fetchAllArticles: fetchAllArticlesData,
    resetError: resetArticlesError
}

ArticleContainer.propTypes = {
    resetError: PropTypes.func.isRequired,
    fetchArticle: PropTypes.func.isRequired,
    fetchAllArticles: PropTypes.func.isRequired,
    article: PropTypes.object.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(React.memo(ArticleContainer));
