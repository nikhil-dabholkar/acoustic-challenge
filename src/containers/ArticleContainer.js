import React from 'react';
import { connect } from 'react-redux';
import { fetchSelectedArticleData, fetchAllArticlesData } from '../actionCreators/ArticleActions';
import { switchLanguagePreference } from '../actionCreators/LanguageActions';
import { BrowserRouter as Router, Route, Switch, Redirect, Link } from "react-router-dom";
import { lazy, Suspense } from 'react';
import { LoadingComponent, LanguageSwitcherComponent } from '../components/index';
import { ErrorBoundaryWrapper } from '../wrappers/index';
import { Breadcrumb, BreadcrumbItem } from 'carbon-components-react';
import styles from './styles/ArticleContainer.module.scss';
import HeaderComponent from '../components/shared/HeaderComponent';
const ListComponent = lazy(() => import("../components/article/ListComponent"));
const DetailsComponent = lazy(() => import("../components/article/DetailsComponent"));
const ErrorComponent = lazy(() => import("../components/shared/ErrorComponent"));

function ArticleContainer({ language, switchLanguage, fetchArticle, fetchAllArticles, article }) {
    return (<React.Fragment>
        <Router>
            <HeaderComponent language={language} switchLanguage={switchLanguage}></HeaderComponent>
            <ErrorBoundaryWrapper>
                <Switch>
                    <Route path="/articles" render={(props) => {
                        return (<Suspense fallback={<LoadingComponent />}>
                            <ListComponent loading={article.loading} error={article.error} allArticles={article.allArticles} fetchAllArticles={fetchAllArticles} {...props} />
                        </Suspense>);
                    }} exact />
                    <Route path="/articles/:id" render={(props) => {
                        return (<Suspense fallback={<LoadingComponent />}>
                            <DetailsComponent loading={article.loading} error={article.error} selectedArticle={article.selectedArticle} fetchArticle={fetchArticle} {...props} />
                        </Suspense>)
                    }} exact />
                    <Route path="/" render={() => <Redirect to="/articles" />} exact></Route>
                    <Route path="*" render={() => {
                        return (<Suspense fallback={<LoadingComponent />}>
                            <ErrorComponent />
                        </Suspense>)
                    }} exact />
                </Switch>
            </ErrorBoundaryWrapper>
        </Router>

    </React.Fragment>
    );
}

const mapStateToProps = state => {
    return ({
        article: state.articleReducer,
        language: state.languageReducer
    })
};

const mapDispatchToProps = {
    fetchArticle: fetchSelectedArticleData,
    fetchAllArticles: fetchAllArticlesData,
    switchLanguage: switchLanguagePreference
}

export default connect(mapStateToProps, mapDispatchToProps)(ArticleContainer);
