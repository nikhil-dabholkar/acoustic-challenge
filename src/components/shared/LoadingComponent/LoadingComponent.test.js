import React from 'react';
import ArticleContainer from '../../../containers/ArticleContainer/ArticleContainer';
import LoadingComponent from './LoadingComponent';
import { IntlProvider } from 'react-intl';
import { store } from "../../../store/store";
import { Provider } from 'react-redux';
import { render, cleanup, screen } from '@testing-library/react';
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import en_US from "../../../i18n/en_US.json";
import { switchLanguagePreference } from "../../../actionCreators/LanguageActions";

afterEach(cleanup);

function renderWithRouter(
    ui,
    { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {}
) {
    return {
        ...render(<Router history={history}>{ui}</Router>),
        history,
    }
}

function rendererWithRouter(
    ui,
    { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {}
) {
    return {
        ...renderer.create(<Router history={history}>{ui}</Router>),
        history,
    }
}

describe('LoadingComponent', () => {
    test('Check if LoadingComponent Renders', async () => {
        const route = '/articles';
        const {findByTestId} = renderWithRouter(<Provider store={store}><IntlProvider locale="en" messages={en_US}><ArticleContainer></ArticleContainer></IntlProvider></Provider>, { route })
        const lazyElement = await findByTestId("loadingComponent")
        expect(lazyElement).toBeInTheDocument()
    })

    test('Check if LoadingComponent Renders',  () => {
        const route = '/articles/fa9519d5-0363-4b8d-8e1f-627d802c08a8';
        const {getByTestId} = renderWithRouter(<Provider store={store}><IntlProvider locale="en" messages={en_US}><ArticleContainer></ArticleContainer></IntlProvider></Provider>, { route })
        const lazyElement =  getByTestId("loadingComponent")
        expect(lazyElement).toBeInTheDocument()
    })

    test('Check if LoadingComponent Renders', async () => {
        const route = '/articles/invalid_item';
        renderWithRouter(<Provider store={store}><IntlProvider locale="en" messages={en_US}><ArticleContainer></ArticleContainer></IntlProvider></Provider>, { route })
        const lazyElement = await screen.findByTestId("loadingComponent")
        expect(lazyElement).toBeInTheDocument()
    })

    it("Check if matches loadingComponent Snapshot", async () => {
        const route = '/articles';
        const tree = rendererWithRouter(<Provider store={store}><IntlProvider locale="en" messages={en_US}><loadingComponent></loadingComponent></IntlProvider></Provider>, { route }).toJSON();
        expect(tree).toMatchSnapshot();
    })
})
