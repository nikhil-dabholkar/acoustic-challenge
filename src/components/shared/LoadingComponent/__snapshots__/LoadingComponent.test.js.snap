// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingComponent Check if matches loadingComponent Snapshot 1`] = `<loadingComponent />`;
